tls/tls_srv.c

```diff
@@ -391,20 +394,29 @@ ttls_parse_alpn_ext(TlsCtx *tls, const unsigned char *buf, size_t len)
 		}
 	}
 
-	/* Use our order of preference. */
-	for (i = 0; i < TTLS_ALPN_PROTOS && alpn_list[i].name; ++i) {
-		our = &alpn_list[i];
-		for (theirs = start; theirs != end; theirs += cur_len) {
-			cur_len = *theirs++;
-			if (ttls_alpn_ext_eq(our, theirs, cur_len)
-			    && ttls_alpn_match_cb(tls, our))
-			{
-				tls->alpn_chosen = our;
-				return 0;
+
+	tls->alpn_chosen = NULL;
+	for (theirs = start; theirs != end; theirs += cur_len) {
+		cur_len = *theirs++;
+		for (i = 0; i < TTLS_ALPN_PROTOS && alpn_list[i].name; ++i) {
+			our = &alpn_list[i];
+			if (ttls_alpn_ext_eq(our, theirs, cur_len)) {
+				/* Calculate JA5t */
+				tls->sess.ja5t.alpn <<= 2;
+				tls->sess.ja5t.alpn += our->id;
+				/* Use our order of preference */
+				if (alpn_chosen_idx > i && ttls_alpn_match_cb(tls, our)) {
+					alpn_chosen_idx = i;
+					tls->alpn_chosen = our;
```

Comment:
It seems there are issues in the logic:

1. we save the last matching ALPN instead of 1st because we don't break loop now on first match
2. if we have `http/1.1,h2` from a client and say `h2,http/1.1` in Tempesta, then we account `http/1.1,h2` twice in `tls->sess.ja5t.alpn`
3. other ALPNs aren't accounted at all, but if a client sends us something non-standard, then it could be quite a signature

Reply:
1. We still choose the 1st matching alpn from our list by saving the least matching index in alpn_chosen_idx, i.e. our order of preference is used.
2. In outer loop we have http/1.1, then iterating in the inner loop we find http/1.1 in our list and we add http/1.1 to ja5t. In the second iteration of outer loop we have h2 and find it in our list and add to ja5t. We use `ttls_alpn_ext_eq` to decide whether to use current alpn in ja5t calculation, so only one calculation takes place for every alpn value

Reply:
3. Added has_unknown_alpn flag in ja5t